pkg/controller/machinehealthcheck/machinehealthcheck_controller.go

```diff
@@ -352,3 +377,71 @@ func isMaster(machine mapiv1.Machine, client client.Client) bool {
 	}
 	return false
 }
+
+// isMachineDisruptionAllowed returns true if the provided MachineDisruptionBudget allows any disruption
+func isMachineDisruptionAllowed(mdb *healthcheckingv1alpha1.MachineDisruptionBudget, maxDisruptedMachinSize int) bool {
+	if mdb.Status.ObservedGeneration < mdb.Generation {
+		glog.Warningf("The machine disruption budget %s is still being processed by the server", mdb.Name)
+		return false
+	}
+	if mdb.Status.MachineDisruptionsAllowed < 0 {
+		glog.Warningf("The machine disruption budget %s MachineDisruptionsAllowed is negative", mdb.Name)
+		return false
+	}
+	if len(mdb.Status.DisruptedMachines) > maxDisruptedMachinSize {
+		glog.Warningf("The machine disruption budget %s DisruptedMachines map too big - too many deletions not confirmed by MDB controller", mdb.Name)
+		return false
+	}
+	if mdb.Status.MachineDisruptionsAllowed == 0 {
+		glog.Warningf("Cannot remediate machine as it would violate the machine's disruption budget %s", mdb.Name)
+		return false
+	}
```

Comment:
I have a cluster with the following nodes:
```
NAME                                   STATUS     ROLES     AGE       VERSION
0bfe7aa3-000c-4f20-b75c-558effd3ecf0   Ready	  worker    3h45m     v1.13.7-beta.0-1
243b4181-1505-4ae2-999b-6b10224840a5   NotReady   worker    11m       v1.13.7-beta.0-1
595dce88-c9ca-484e-a44d-c5919f5b7868   NotReady   worker    11m       v1.13.7-beta.0-1
63141c47-5892-4ea2-ba09-c9c976831689   NotReady   worker    11m       v1.13.7-beta.0-1
80aedfb8-3c2a-43fe-9f65-9c9219a13e07   NotReady   worker    11m       v1.13.7-beta.0-1
8fa00196-9d60-4496-9373-df2743e10b89   NotReady   worker    11m       v1.13.7-beta.0-1
a17d1e3f-b632-43c4-99b3-3fc8c30f73e7   NotReady   worker    11m       v1.13.7-beta.0-1
e158da13-8a14-41bc-a4c0-171254b40ec0   NotReady   worker    11m       v1.13.7-beta.0-1
fcf4ebbd-6bb3-4716-8175-2af362d8dfe3   NotReady   worker    11m       v1.13.7-beta.0-1
fd6bdee7-1057-4134-884c-12d58746a733   NotReady   worker    11m       v1.13.7-beta.0-1
ip-10-0-101-95.ec2.internal            Ready	  master    4h10m     v1.13.1
```

MHC and MDB (with `maxUnavailable` set to 1) are deployed. MHC is continuously reporting:
```
W0712 15:02:26.733652   14737 machinehealthcheck_controller.go:385] cannot delete machine as it would violate the machine's disruption budget mdb-test
W0712 15:02:26.733675   14737 machinehealthcheck_controller.go:224] mahcine disruption is not allowed
```

Given the disruption budget is already violated even before MHC starts to remediate, none of the NotReady nodes will ever get healed. Is that the expected behavior? I would assume at least one node at a time will be allowed to be  healed.

Reply:
Ok, I get it now. I had different assumption about how the disruption budget is used. There are two perspectives actually:
1) Cluster is assumed to be healthy. There is upper bound to a number of node that can be unhealthy so in case the bound is set e.g. to 4 and at most 4 nodes switch to NotReady state, MHC controller can remediate those nodes.
2) In case I create the disruption budget for a cluster where too many nodes are unhealthy, I expect the MHC controller to do absolutely nothing about it.

So, this behavior is actually correct.

Reply:
My original though was another scenario that is currently not handled by the disruption budget. In case there are more nodes unhealthy than the disruption budget allows I was expecting the MHC controller to select a subset of at most `maxUnavailable` nodes to heal (so it would kinda respect the fact that only at most `maxUnavailable` machines are healed). The confusion hear is semantics of the word disruption. My first impression was it means at most `maxUnavailable` nodes can be healed. Not the fact, that if a set of nodes has more than `maxUnavailable` nodes unhealthy, entire set of the nodes is considered too fragile to be touched.